### ISSUE TITLE
Allocations incurred from broadcast expression

### ISSUE THREAD

**Issue by charleskawczynski** [CliMA/ClimaCore.jl#1372]
[This line](https://github.com/CliMA/ClimaAtmos.jl/blob/021a121cfd0c38152dc5f78d509082bec9ebe768/src/parameterized_tendencies/microphysics/precipitation.jl#L97-L98) is breaking on the GPU, and is resulting in runtime allocations.

I'll try to make a MWE when I have a chance.

**Comment by simonbyrne** [comment]
The simplest example is when you broadcast a `DataType` over a scalar, which is then combined with the broadcast over another object, e.g.
```julia
FT = Float32
@. x + FT(1)
```
which is equivalent to
```julia
x .+ FT.(1)
```

Note that this is distinct from https://github.com/JuliaLang/julia/issues/50554 (which is due to the use of `flatten`). You can hit this with plain `CuArray`s, see https://github.com/JuliaGPU/CUDA.jl/issues/1761.

Unfortunately this is a common pattern we use to make functions type-generic. Some possible solutions:

The first is simply to avoid broadcasting types over scalars. We could either switch away from using `@.`, so the above would be
```julia
x .+ FT(1)
```
which forces the the `Float32(1)` to be evaluated before broadcasting. You can use `$()` to do the same with `@.`:
```julia
@. x + $(FT(1))
```
but this is kind of ugly. Finally you can just move the conversion outside
```julia
i = FT(1)
@. x + i
```

If we want to keep using it, one option is via type-piracy on `Base.Broadcast.broadcasted`, a la https://github.com/CliMA/ClimaCore.jl/pull/1365, but I would prefer not to do this. I think the suggestion in https://github.com/JuliaGPU/CUDA.jl/issues/1761#issuecomment-1425496244 of using `Adapt` is probably the best option, but it would have to be done in CUDA.jl or GPUArrays.jl.

Perhaps a more elegant solution is
```julia
struct SuffixConverter{FT}
end
Base.:*(x::Number, ::SuffixConverter{FT}) where {FT} = convert(FT, x)
Base.Broadcast.broadcasted(::typeof(*), x::Number, ::SuffixConverter{FT}) where {FT} = convert(FT, x)
```
Then we can write
```julia
_FT = SuffixConverter{FT}()
@. x + 1_FT
```
which has the nice advantage of avoiding more parentheses, and so could make the code cleaner.

**Comment by simonbyrne** [comment]
I've posted a patch to CUDA.jl which should address it: https://github.com/JuliaGPU/CUDA.jl/pull/2000

**Comment by charleskawczynski** [comment]
I like the `_FT` suffix, it reminds me of Fortran syntax, but it can work locally with the given generic float type. It saves one character (`x_FT` instead of `FT(x)`) and it's relatively simple. My only question is: Where does this definition live? Is this something that ClimaCore will define? It's kind of low level and doesn't really have anything to do with spatial operators.

**Comment by simonbyrne** [comment]
Okay, how about https://github.com/simonbyrne/SuffixConversion.jl

**Comment by charleskawczynski** [comment]
I think that’s a great solution, and probably better that it lives outside of clima since it’s very widely applicable.

**Comment by milankl** [comment]
I often end up writing things like
```julia
half = convert(NF,0.5)
dt_NF = convert(NF,dt)
Tmin = convert(NF,scheme.Tmin)
vj = convert(NF,v[j])
```
(SpeedyWeather.jl uses NF for number format) I like `convert` as it's more verbose that we're just converting a variable, but don't actually do any computation. Someone not being used to `FT(1)` may think you're doing something much more sophisticated here? But honestly, I often just end up unpacking/converthing things in one go, like `Tmin = convert(NF,scheme.Tmin)` and I still very much prefer that over the conversion in the actual kernel. I'd rather be explicit than relying on the compiler to always treat `NF(1)` as a constant instead of redoing the conversion over and over again.

**Comment by Sbozzolo** [comment]
https://github.com/JuliaGPU/CUDA.jl/pull/2000 got merged, so maybe this is fixed?

**Comment by charleskawczynski** [comment]
This closed for 1.11, but is still an issue on 1.10.
